lib/sshutils/ctx.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sshutils
+
+import (
+	"io"
+	"net"
+	"sync"
+
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+
+	"github.com/gravitational/trace"
+)
+
+// ConnectionContext manages connection-level state.
+type ConnectionContext struct {
+	sync.RWMutex
```

Comment:
Don't embed the mutex, nothing outside of this type should be able to mess with it.

Also, keep the mutex above the fields it protects, as a hint to the reader. In this case, `Connection` and `ServerConn` are not guarded by the mutex, so they should be above.